python/graphstorm/trainer/glem_np_trainer.py

```diff
@@ -160,25 +161,42 @@ def _fit_one_epoch(self, use_gnn, model, g, data, train_loader,
                        save_model_frequency=-1, no_pl=False):
         """Fit model for one epoch
         """
-        profiler.start_record()
-        for i, (input_nodes, seeds, blocks) in enumerate(train_loader):
-            profiler.record('train_sample')
-            total_steps += 1
-            batch_tic = time.time()
-
+        def _prepare_batch(input_nodes, seeds, blocks, is_labeled=True):
+            """Prepare a batch of graph data from the data loader, by retrieving features,
+            moving blocks to device, and get labels if `is_labeled` is True.
+            """
             if not isinstance(input_nodes, dict):
                 assert len(g.ntypes) == 1
                 input_nodes = {g.ntypes[0]: input_nodes}
+
             input_feats = data.get_node_feats(input_nodes, device)
-            lbl = data.get_labels(seeds, device)
             profiler.record('train_node_feats')
-
+            lbl = None
+            if is_labeled:
+                lbl = data.get_labels(seeds, device)
             blocks = [block.to(device) for block in blocks]
-            for _, feats in input_feats.items():
-                num_input_nodes += feats.shape[0]
             profiler.record('train_graph2GPU')
+            return input_nodes, input_feats, blocks, lbl
+
+        profiler.start_record()
+        for i, batch in enumerate(train_loader):
+            if isinstance(train_loader, GSgnnNodeSemiSupDataLoader):
+                # semi-supervised setting
+                input_nodes, input_feats, blocks, lbl = _prepare_batch(*batch[0])
+                input_nodes_u, input_feats_u, blocks_u, _ = _prepare_batch(
+                    *batch[1], is_labeled=False)
+            else:
+                # supervised setting, no unlabeled data
+                input_nodes, input_feats, blocks, lbl = _prepare_batch(*batch)
+                input_nodes_u = input_feats_u = blocks_u = None
+            profiler.record('train_sample')
+            total_steps += 1
+            batch_tic = time.time()
             # Run forward function to compute loss:
-            loss = model(blocks, input_feats, None, lbl, input_nodes, use_gnn=use_gnn, no_pl=no_pl)
+            loss = model(blocks, input_feats, None, lbl, input_nodes, use_gnn=use_gnn,
+                         no_pl=no_pl or (epoch < model.module.num_pretrain_epochs),
```

Comment:
should this argument be associated with the model? not with the trainer?

Reply:
It can be associated with either. But this is specific to GLEM, currently GLEM trainer doesn't have configurable arguments. So I think it's clearer if all options for training GLEM are grouped to the model like this: 

```yml
  lmgnn:
    lm_train_nodes: 10
    lm_infer_batch_size: 32
    freeze_lm_encoder_epochs: 0
    model_encoder_type: rgcn
    fanout: "4"
    num_layers: 1
    hidden_size: 128
    use_mini_batch_infer: true
    training_method:
      name: glem
      kwargs:
        em_order_gnn_first: false
        inference_using_gnn: true
        pl_weight: 0.5
        num_pretrain_epochs: 1
```

Reply:
i think so. what i mean is whether we should attach `num_pretrain_epochs` to the model object or the trainer object.